internal/e2e/artifact_test.go

```diff
@@ -0,0 +1,35 @@
+//go:build e2e
+
+package e2e
+
+import (
+	"strings"
+	"testing"
+)
+
+func TestArtifactUploadDownload(t *testing.T) {
+	ctx := t.Context()
+
+	tc := newTestCase(t, `
+agents:
+  queue: {{.queue}}
+steps:
+  - commands:
+    - echo "hello world" > artifact.txt
+    - buildkite-agent artifact upload artifact.txt
+  - commands:
+    - buildkite-agent artifact download artifact.txt . && if [[ $(cat artifact.txt) == "hello world" ]]; then echo "success" && exit 0; else echo "failure" && exit 1; fi
+`)
+
+	tc.startAgent()
+	build := tc.triggerBuild()
+	state := tc.waitForBuild(ctx, build)
+	if got, want := state, "passed"; got != want {
+		t.Errorf("Build state = %q, want %q", got, want)
+	}
+
+	logs := tc.fetchLogs(ctx, build)
+	if !strings.Contains(logs, "\nsuccess") {
+		t.Errorf("tc.fetchLogs(ctx, build %q) logs as follows, did not contain 'success'\n%s", build.ID, logs)
+	}
```

Comment:
Relying on the job to succeed or fail might be enough. Having to pull the job logs might make the test run longer for not much benefit.